cpmpy/solvers/z3.py

```diff
@@ -368,7 +392,23 @@ def _z3_expr(self, cpm_con, reify=False):
                     return z3.And(self._z3_expr(any(lhs == a for a in rhs.args)),
                                   self._z3_expr(all([lhs <= a for a in rhs.args])))
 
-                lhs, rhs = self._z3_expr(cpm_con.args)
+
+                #comparing a boolexpr with an integer or float:
+                if is_num(rhs) and not is_bool(rhs) and lhs.is_bool():
+                    iv = _IntVarImpl(0, 1)  # 0,1 is the valid domain, as this represents true or false
+                    mhs = self._z3_expr(iv)  # turn into z3 expression
+                    lhs, rhs = self._z3_expr(cpm_con.args)
+                    if is_any_list(lhs):
+                        olhs = lhs[0]
+                        orhs = lhs[1]
+                        #keep auxiliary assignments in the first element of the list
+                        return [z3.And(olhs, orhs == mhs), mhs == rhs]
```

Comment:
nonono, you can not sometimes return a list (tuple where first element is special wrt second/others) and sometimes a single element.
- python allows it but you should avoid doing it
- especially on a function that is called in so many places!
- if you must, you should always return a tuple with a first special element and then second element a potentially empty list (like in flatten)

but here you really shouldn't...

A design principle of CPMpy is that the user can create any type of expression, and a solver should only accept the expressions that it can straightforwardly map to its internal API.

The heavy lifting of going from any expression to those supported is put into transformations, such that other (future) solvers with similar needs can reuse the same transformations...

So here, you should create a new transformation file 'strict_bool' or whathever is more appropriate, and create these auxiliary channeling variables in there, at CPMpy expression level.

Then the z3 interface can just post the cpmpy expressions as before (with previously unsupported cases now transformed to expressions (with new vars) that it does support)

Reply:
I added it as a transformation in comparison.py
But actually @JoD thought it might not be needed after all, and he was correct.
Where needed boolrefs are now cast to int as follows: if boolref then 1 else 0

Don't know if the transformation will be useful in the future so maybe we should not keep it?

In some cases (comparing 2 boolrefs with >,<,<= or >=) it might be more efficient to use implies, can still implement this if it seems better to you. ( b >= c is equivalent to c.implies(b) ) 

I also added a function is_boolref to utils, to avoid having to check for the 'is_bool' attribute every time.
This won't be needed if we use fishhook in the future but I think it's useful for now.